misc/python/materialize/cli/mzbench.py

```diff
@@ -0,0 +1,160 @@
+        composition,
+        "run",
+        f"setup-{benchmark}",
+    ]
+    run_benchmark = [
+        "./bin/mzcompose",
+        "--mz-find",
+        composition,
+        "run",
+        f"run-{benchmark}",
+    ]
+
+    field_names = [
+        "git_revision",
+        "num_workers",
+        "seconds_taken",
+        "rows_per_second",
+        "grafana_url",
+    ]
+    results_writer = csv.DictWriter(sys.stdout, field_names)
+    results_writer.writeheader()
+
+    # We use check_output because check_call does not capture output
+    try:
+        subprocess.check_output(setup_benchmark, stderr=subprocess.STDOUT)
+    except (subprocess.CalledProcessError,) as e:
+        print(f"Setup benchmark failed! Output from failed command:\n{e.output}")
+        raise
+
+    for (worker_count, git_revision) in itertools.product(worker_counts, git_revisions):
+
+        child_env = os.environ.copy()
+        child_env["MZ_WORKERS"] = str(worker_count)
+        child_env["MZBUILD_WAIT_FOR_IMAGE"] = "true"
+        if git_revision:
+            child_env["MZBUILD_MATERIALIZED_TAG"] = f"unstable-{git_revision}"
+
+        try:
+            output = subprocess.check_output(
+                run_benchmark, env=child_env, stderr=subprocess.STDOUT
+            )
+        except (subprocess.CalledProcessError,) as e:
+            print(f"Setup benchmark failed! Output from failed command:\n{e.output}")
+            raise
+
+        # TODO: Replace parsing output from mzcompose with reading from a well known file or topic
+        for line in output.decode().splitlines():
+            if line.startswith("SUCCESS!"):
+                for token in line.split(" "):
+                    if token.startswith("seconds_taken="):
+                        seconds_taken = token[len("seconds_taken=") :]
+                    elif token.startswith("rows_per_sec="):
+                        rows_per_second = token[len("rows_per_sec=") :]
+            elif line.startswith("Grafana URL: "):
+                grafana_url = line[len("Grafana URL: ") :]
+
+        results_writer.writerow(
+            {
+                "git_revision": git_revision if git_revision else "NONE",
+                "num_workers": worker_count,
+                "seconds_taken": seconds_taken,
+                "rows_per_second": rows_per_second,
+                "grafana_url": grafana_url,
+            }
+        )
+
+
+def enumerate_cpu_counts() -> typing.List[int]:
+    """This program prints the number of CPU counts to benchmark on this machine.
+
+    We remove some percentage of CPU cores off the top for system / background processing. With
+    the CPUs that remain, we generate a list of evenly spaced worker counts. The list is limited
+    by the number of trials desired. This is meant to help us explore the number of CPUs that
+    should be dedicated to MZ_WORKERS, not as a prescription for the correct values to choose.
+
+    On a Macbook with 8 cores, this will return [3, 2, 1].
```

Comment:
I'm interpreting this comment as "on a 8 core macbook, we'll try benchmarking with mz_workers = 3, then 2 then 1" which seems a bit overly conservative if there are 8 proper cores? Running with 6 workers works totally fine on my laptop. I could also be misinterpreting the comment entirely

Reply:
Sorry, I think this comment may be out of date. The comment was supposed to say "with 8 processors" where processors may be logical or physical. Now that we're using `psutil` to determine the number of processors, I'll update the comment. And yeah, we should use 6 workers if you have 8 *physical* cores.